dist/centrifuge.d.ts

```diff
@@ -0,0 +1,115 @@
+// Type definitions for centrifuge 2.*.*
+// Project: https://github.com/centrifugal/centrifuge-js
+// Definitions by: Jekaspekas <https://github.com/jekaspekas>
+// TypeScript Version: 3.4.5
+
+import EventEmitter from 'events';
+
+export = Centrifuge;
+
+declare class Centrifuge extends EventEmitter {
+    constructor(url: string, options?: Centrifuge.Options);
+    setToken(token: string): void;
+    setConnectData(data: any): void;
+    rpc(data: any): Promise<any>;
+    send(data: any): Promise<{}>;
+    publish(channel: string, data: any): Promise<{}>;
+    getSub(channel: string): Centrifuge.Subscription;
+    isConnected(): boolean;
+    connect(): void;
+    disconnect(): void;
+    ping(): void;
+    startBatching(): void;
+    stopBatching(): void;
+    startSubscribeBatching(): void;
+    stopSubscribeBatching(): void;
+    subscribe(channel: string, events?: (...args: any[]) => void): Centrifuge.Subscription;
+    subscribe(channel: string, events?: Centrifuge.SubscriptionEvents): Centrifuge.Subscription;
+}
+
+declare namespace Centrifuge {
+    export interface Options {
+        debug?: boolean;
+        sockjs?: any;
+        promise?: any;
+        minRetry?: number;
+        maxRetry?: number;
+        timeout?: number;
+        ping?: boolean;
+        pingInterval?: number;
+        pongWaitTimeout?: number;
+        privateChannelPrefix?: string;
+        onTransportClose?: boolean;
+        sockjsServer?: string | null;
+        sockjsTransports?: string[];
+        refreshEndpoint?: string;
+        refreshHeaders?: object;
+        refreshParams?: object;
+        refreshData?: object;
+        refreshAttempts?: number | null;
+        refreshInterval?: number;
+        onRefreshFailed?: () => void;
+        onRefresh?: (context: object, cb: (resp: any) => void) => void;
+        subscribeEndpoint?: string;
+        subscribeHeaders?: object;
+        subscribeParams?: object;
+        subRefreshInterval?: number;
+        onPrivateSubscribe?: (message: {data: SubscribePrivateContext}, cb: (resp: any) => void) => void;
+    }
+    export class Subscription extends EventEmitter {
+        channel: string;
+        ready(callback: (context: SubscribeSuccessContext) => void, errback: (context: SubscribeErrorContext) => void): void;
+        subscribe(): void;
+        unsubscribe(): void;
+        publish(data: any): Promise<any>;
```

Comment:
There is a possibility to declare `Promise.then()` function argument.
To do this, specify the type in `Promise<>`.
But Subscription.publish(), Subscription.presence(), Subscription.presenceStats(), Subscription.history() transmit data in different formats for Protobuf and JSON.

Reply:
What do you mean under different formats? Do you mean different type of objects returned here? At least those objects contain the same fields right? So we can just create wrapper objects like we just did for `PublishMessage`.

Reply:
Subscription.presenceStats(), Subscription.history() methods. CentrifugeProtobuf return success callback message at next format:
> ![protobuf](https://user-images.githubusercontent.com/8606056/58382538-af85b000-7ff5-11e9-84f0-dfe275284357.jpg)
> it is different from the format returned Centrifuge:
> ![json](https://user-images.githubusercontent.com/8606056/58382552-d9d76d80-7ff5-11e9-9723-9260aef0ba90.jpg)

I think wrapper need here

Reply:
Hmm, it's a bit surprising to me that Protobuf uses camelCase even though that fields defined in snake_case in proto file. I will try to check whether it's possible to come to the same field names. Do we still need wrappers even if fields will have the same names? Can't we just describe both cases using sth like:

```json
    export interface ClientInfo {
        user? : string;
        client? : string;
        default_info?: any;
        channel_info?: any;
    }

    export interface PresenceResult {
        presence: PresenceMap;
    }

    export interface PresenceMap {
        [key: string]: ClientInfo}
    }
```
?

Reply:
Then it is necessary to declare two types: ClientInfo & ClientInfoProtobuf
```ts
    export interface PresenceMap {
        [key: string]: ClientInfo | ClientInfoProtobuf
    }
```
And check type in promise then for get object fields

Reply:
Can we discuss this in Telegram btw (will be much simpler for me to understand)?

Reply:
https://github.com/centrifugal/centrifuge-js/commit/0b8e65df0ef6ed3f959233f99ae0cf1463560a6a - this commit fixes different names

Reply:
In order to better understand you need to start using Typescript.
For example i set `presence(): Promise<PublishContext>;` (i know PublishContext is wrong type for this):
![example](https://user-images.githubusercontent.com/8606056/58482502-b1cc4380-8188-11e9-83c8-d6ecd0d6aa2e.jpg)

On the right, this is how the user will use it.

Reply:
I can not understand what format to use for the declaration:
 - Documentation (have channel and data top-level fields):
```json
{
    "channel":"$public:chat",
    "data":{
        "2724adea-6e9b-460b-4430-a9f999e94c36": {
            "user":"2694",
            "client":"2724adea-6e9b-460b-4430-a9f999e94c36"
        },
        "d274505c-ce63-4e24-77cf-971fd8a59f00":{
            "user":"2694",
            "client":"d274505c-ce63-4e24-77cf-971fd8a59f00"
        }
    }
}
```
 - Protobuf (no channel field, presence instead data field, add chanInfo and connInfo):
```json
{
    "presence":{
        "2724adea-6e9b-460b-4430-a9f999e94c36": {
            "chanInfo": Uint8Array [],
            "user":"2694",
            "connInfo": Uint8Array [],
            "client":"2724adea-6e9b-460b-4430-a9f999e94c36"
        },
    }
}
```
 - JSON (no channel field, presence instead data field, no chanInfo and connInfo):
```json
{
    "presence":{
        "2724adea-6e9b-460b-4430-a9f999e94c36": {
            "user":"2694",
            "client":"2724adea-6e9b-460b-4430-a9f999e94c36"
        },
    }
}
```

Reply:
Documentation is wrong, this is left after migration from v1. I'll fix it.

The source of truth is [proto file](https://github.com/centrifugal/centrifuge/blob/master/misc/proto/client.proto). Now with fix from https://github.com/centrifugal/centrifuge-js/commit/0b8e65df0ef6ed3f959233f99ae0cf1463560a6a I suppose both formats match the same schema.

Reply:
For Protobuf proto3 syntax all fields are optional. It is difficult for me to determine which fields are required and which are optional for typescript declarations, so you need to check interfaces:
- PresenceResult
- PresenceStatsResult
- HistoryResult

Reply:
conn_info and chan_info - are optional fields (? symbol) for type definition.

Reply:
Yeah, I got the main idea. There are still several problems with names and field types - but suppose I can fix them myself, and need to test it out locally. One more question is about Promises - Promise contains callback and errback functions, here we only define callback parts - could you explain why you omitted errbacks? Is it convenient for Typescript declarations or you just don't have enough info?

Reply:
[Errback type is hardcoded to `any`](https://github.com/Microsoft/TypeScript/issues/7588)